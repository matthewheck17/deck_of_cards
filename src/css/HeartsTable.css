#game-container {
    height: auto;
    width: auto;
}

#start-menu {
    position: absolute;
    top: 0;
    height: 100%;
    width: 100%;
    background: rgba(0,0,0,0.3);
    z-index: 999;
}

.start-menu-text {
    margin: 0;
    text-align: center;
    font-family: "Andale Mono", AndaleMono, monospace;
    font-size: 75px;
    color: white;
    padding-top: 50px;
}

.start-menu-text:hover {
    color: rgb(211, 217, 223);
    font-size: 78px;
}

#begin-button {
    margin-top: 2em;
}

#mockDeck {
    z-index: 998;
}

#playing-area {
    position: absolute;
    height: 175px;
    width: 175px;
    top: 50%;
    margin-top: -87.5px;
    left: 50%;
    margin-left: -87.5px;
    border: 2px solid #2b5440;
    border-radius: 5px;
}

#hearts-logo {
    position: absolute;
    height: 50px;
    width: 50px;
    top: 50%;
    margin-top: -25px;
    left: 50%;
    margin-left: -25px;
}

.player-hand {
    position: absolute;
}

#player1-hand {
    bottom: 0;
    left: 50%;
    margin-left: -250px;
    border: 2px solid #2b5440;
    border-radius: 5px;
    height: 225px;
    margin-bottom: -100px;
    width: 500px;
}

#player2-hand {
    left: 0;
    top: 50%;
    margin-top: -200px;
    height: 400px;
    width: 225px;
    margin-left: -100px;
    border: 2px solid #2b5440;
    border-radius: 5px;
}

#player3-hand {
    top: 0;
    left: 50%;
    margin-left: -250px;
    border: 2px solid #2b5440;
    border-radius: 5px;
    height: 225px;
    margin-top: -100px;
    width: 500px;
}

#player4-hand {
    right: 0;
    top: 50%;
    margin-top: -200px;
    height: 400px;
    width: 225px;
    margin-right: -100px;
    border: 2px solid #2b5440;
    border-radius: 5px;
}