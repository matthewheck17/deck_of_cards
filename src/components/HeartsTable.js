// Cpoyright Â© 2020 Matthew Heck. All rights reserved.

// This file contains all necessart items for the HeartsTable Component

// Created in US 1.1

// SYSTEM IMPORTS
import React from "react";

// CUSTOM IMPORTS
import GameCard from "./GameCard.js";
import "../css/HeartsTable.css";
import HeartsLogo from "../images/hearts/hearts-logo.png";

class HeartsTable extends React.Component {

  constructor(props) {
    super(props)
    // array holding all cards
    var allCards = [
      ["club","A"],["club","2"],["club","3"],["club","4"],["club","5"],["club","6"],["club","7"],["club","8"],["club","9"],["club","10"],["club","Jack"],["club","Queen"],["club","King"],
      ["diamond","A"],["diamond","2"],["diamond","3"],["diamond","4"],["diamond","5"],["diamond","6"],["diamond","7"],["diamond","8"],["diamond","9"],["diamond","10"],["diamond","Jack"],["diamond","Queen"],["diamond","King"],
      ["spade","A"],["spade","2"],["spade","3"],["spade","4"],["spade","5"],["spade","6"],["spade","7"],["spade","8"],["spade","9"],["spade","10"],["spade","Jack"],["spade","Queen"],["spade","King"],
      ["heart","A"],["heart","2"],["heart","3"],["heart","4"],["heart","5"],["heart","6"],["heart","7"],["heart","8"],["heart","9"],["heart","10"],["heart","Jack"],["heart","Queen"],["heart","King"]
    ];

    // 4 empty arrays that will hold player hands
    var hand1 = [];
    var hand2 = [];
    var hand3 = [];
    var hand4 = [];
    const handSize = 13; //constant holds the number of cards in a hand

    // loop iterates the number of times that the handsize is
    for(var index = 0; index < handSize; index++){
      var rand = Math.floor(Math.random() * (allCards.length)); //get a random card from the allCards array
      hand1.push(allCards[rand]);  //push it to hand1
      allCards.splice(rand, 1);  //remove it from the allCards array
      rand = Math.floor(Math.random() * (allCards.length)); //repeat the process for the other three hands
      hand2.push(allCards[rand]);
      allCards.splice(rand, 1);
      rand = Math.floor(Math.random() * (allCards.length));
      hand3.push(allCards[rand]);
      allCards.splice(rand, 1);
      rand = Math.floor(Math.random() * (allCards.length));
      hand4.push(allCards[rand]);
      allCards.splice(rand, 1);
    }

    //set the state of the component
    this.state = {
      hand1: hand1,
      hand2: hand2,
      hand3: hand3,
      hand4: hand4,
      handSize: 13,
      img: this.props.img,
      menu: "showing"
    }
  }

  //function to exit the menu
  exitMenu = () => {
    this.setState({
      menu: "notShowing"
    })
  }

  deal(){
    // Initialize empty deck array
    let cards = [];

    // 4 for loops to initialize each game card component with the card suit and value taken from its respective player hand array
    for (let i=0; i < this.state.hand1.length; i++) {
      cards.push(<GameCard ref={'card'+i} key={i} suit={this.state.hand1[i][0]} value={this.state.hand1[i][1]} img={this.state.img} side="front" location="hand1"/>)
    }

    for (let i=0; i < this.state.hand2.length; i++) {
      cards.push(<GameCard ref={'card'+i+this.state.handSize-1} key={i+this.state.handSize-1} suit={this.state.hand1[i][0]} value={this.state.hand1[i][1]} img={this.state.img} side="back" location="hand2"/>)
    }

    for (let i=0; i < this.state.hand3.length; i++) {
      cards.push(<GameCard ref={'card'+i+(2*this.state.handSize)-1} key={i+(2*this.state.handSize)-1} suit={this.state.hand1[i][0]} value={this.state.hand1[i][1]} img={this.state.img} side="back" location="hand3"/>)
    }

    for (let i=0; i < this.state.hand4.length; i++) {
      cards.push(<GameCard ref={'card'+i+(3*this.state.handSize)-1} key={i+(3*this.state.handSize)-1} suit={this.state.hand1[i][0]} value={this.state.hand1[i][1]} img={this.state.img} side="back" location="hand4"/>)
    }

    return cards;
  }



  render() {
    return (
      <div id='game-container'>
        {/* Menu */}
        {this.state.menu === "showing" &&
          <div>
            <GameCard ref={"mock-deck"} key={"mock-deck"} suit={"spade"} value={"Queen"} img={this.state.img} side="back" location="deck"/>
            <div id='start-menu'>
              <h2 id="begin-button" className="start-menu-text" onClick={this.exitMenu}>Begin</h2>
            </div>
          </div>
        }
        <div id="playing-area">
          <img id="hearts-logo" src={HeartsLogo} href="hearts-logo" alt="heart"/>
        </div>

        <div id="player1-hand" className="player-hand"></div>

        <div id="player2-hand" className="player-hand"></div>

        <div id="player3-hand" className="player-hand"></div>

        <div id="player4-hand" className="player-hand"></div>
        {this.state.menu === "notShowing" &&
          <div>
            {this.deal()}
          </div>
        }
      </div>
    );
  }
}

export default HeartsTable;
