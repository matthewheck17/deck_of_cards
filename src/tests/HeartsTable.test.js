import React from "react";
import {render, cleanup} from "@testing-library/react";
import HeartsTable from "../components/HeartsTable.js";
import '@testing-library/jest-dom';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, configure } from 'enzyme';

//cleanup after each test is ran to unmount react trees mounted with render
afterEach(cleanup);

/* SET OF TESTS FOR FIRST DEVELOPMENT CYCLE */
describe('CYCLE 1 TESTS', () => {

  it ("US 1.1 - Testing that the Hearts Table component renders correctly", () => {
    const {asFragment} = render(<HeartsTable/>) // render a HeartsTable component
    expect(asFragment()).toMatchSnapshot()  //Check that it matches snapshot
  })
  
  configure({adapter: new Adapter()});
  it("US 1.1 - Testing that the start menu shows up and renders with the correct html elements", () => {
    const wrapper = shallow(<HeartsTable/>);
    expect(wrapper.find("div").length).toEqual(8);
  })

  it("US 1.2 - Testing that the start menu goes away after clicking on 'Begin'", () => {
    const wrapper = shallow(<HeartsTable/>);
    wrapper.find('h2').simulate('click');
    expect(wrapper.find("div").length).toEqual(7);
  })
  
  it("US 1.2 - Testing that player 1 is dealt a hand of the 13 cards", () => {
    const wrapper = shallow(<HeartsTable/>);
    var instance = wrapper.instance();
    var cardsInHand = instance.state.hand1.length
    expect(cardsInHand).toEqual(13);
  })

  it("US 1.2 - Testing that player 2 is dealt a hand of the 13 cards", () => {
    const wrapper = shallow(<HeartsTable/>);
    var instance = wrapper.instance();
    var cardsInHand = instance.state.hand2.length
    expect(cardsInHand).toEqual(13);
  })

  it("US 1.2 - Testing that player 3 is dealt a hand of the 13 cards", () => {
    const wrapper = shallow(<HeartsTable/>);
    var instance = wrapper.instance();
    var cardsInHand = instance.state.hand3.length
    expect(cardsInHand).toEqual(13);
  })

  it("US 1.2 - Testing that player 4 is dealt a hand of the 13 cards", () => {
    const wrapper = shallow(<HeartsTable/>);
    var instance = wrapper.instance();
    var cardsInHand = instance.state.hand4.length
    expect(cardsInHand).toEqual(13);
  })
})
