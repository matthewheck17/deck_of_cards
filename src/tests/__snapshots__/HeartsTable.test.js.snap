// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CYCLE 1 TESTS US 1.1 - Testing that the Hearts Table component renders correctly 1`] = `
<DocumentFragment>
  <div
    id="game-container"
  >
    <div>
      <div>
        <div
          class="deck"
          id="back-of-card"
        >
          <img
            alt="card back"
            draggable="false"
            id="back-image"
          />
        </div>
      </div>
      <div
        id="start-menu"
      >
        <h2
          class="start-menu-text"
          id="begin-button"
        >
          Begin
        </h2>
      </div>
    </div>
    <div
      id="playing-area"
    >
      <img
        alt="heart"
        href="hearts-logo"
        id="hearts-logo"
        src="hearts-logo.png"
      />
    </div>
    <div
      class="player-hand"
      id="player1-hand"
    />
    <div
      class="player-hand"
      id="player2-hand"
    />
    <div
      class="player-hand"
      id="player3-hand"
    />
    <div
      class="player-hand"
      id="player4-hand"
    />
  </div>
</DocumentFragment>
`;
